### PR TITLE
Deleting the connection owner? Think again (or transfer ownership.)

### PR DESCRIPTION
**Pull request by dereksmart**
#### Changes proposed in this Pull Request:

- Adds some new helper methods for getting info about the user connections.
- Modifies the permission callback for the `/connection/owner` endpoint.
- Adds a whole notice UI and logic during the delete user process with information and actions to take if the connection owner is being deleted.

**What this PR is intentionally not doing, and will be done in other PRs:**
- Add tracking
- Better copy/design

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* I wouldn't really call it a feature, more a preventative enhancement. 

#### Testing instructions:

There's a lot. 

**Test changing connection owner to secondary Admin**
1. Set up a new site. Connect Jetpack to whatever user - let's call it `User A`. 
2. Add a new admin user to the site - let'a call it `User B`.
3. In a separate (incognito?) window, log into the site as User B. 
4. Go to the Jetpack admin and connect User B to wordpress.com. 
5. Still logged in as User B, go to the users page. 
6. Click to delete User A. (Remember, User A should be the connection owner). You should be taken to a confirmation page to delete, where you should see this new notice: 
image
7. Since you are logged in as a secondary connected user, you should see a dropdown with your name in it. 
8. Click the button to switch the connection to you! 
9. The notice should disappear. 

**Verify it worked**
After you did the above, you should be able to verify it "worked" by doing the following things: 
- In CLI: `wp jetpack options get master_user` - should be User B's ID returned.
- In wordpress.com NA: Visit the blog's report card. You should see User B's wpcom connected user as the new owner. 
- You should also see the connection owner has been updated in the blog's audit trail in wpcom. 

**Test when there are no other users connected**
Since we're only able to transfer ownership to another connected user, we needed a case to handle if there were none! So, you should expect to see a lot of text and a button to connect if this is the case when you're trying to delete the connection owner.  For this one, please verify: 
- That connection link works, and is done in a new tab. 
- The support links work and points to the right places. 
image

**General things to check/verify**
- You should never see an option to transfer ownership to a non-admin
- You should never see an option to transfer ownership to another user that is not connected currently. 

#### Proposed changelog entry for your changes:

*General: Warn users that deleting the connection owner is going to break things.

### CODE REVIEW

File: packages/connection/src/Manager.php
Comment (dereksmart) [inline packages/connection/src/Manager.php]: Happy to break this out into it's own PR if it makes it easier.

File: packages/jitm/src/JITM.php
Comment (dereksmart) [inline packages/jitm/src/JITM.php]: help :)

File: tests/php/_inc/lib/test_class.rest-api-endpoints.php
Comment (dereksmart) [inline tests/php/_inc/lib/test_class.rest-api-endpoints.php]: This was the wrong user cap to allow switching connection owner. Any user can connect_user, but not any user can own the connection.

File: packages/connection/src/Manager.php
Comment (dereksmart) [inline packages/connection/src/Manager.php]: If the user is missing, this will break. I will update this to do the same as it's done in https://github.com/Automattic/jetpack/pull/13323 by returning ID only once it's merged.

File: tests/php/_inc/lib/test_class.rest-api-endpoints.php
Comment (dereksmart) [inline tests/php/_inc/lib/test_class.rest-api-endpoints.php]: The method now allows non-master users to use this endpoint.

File: packages/jitm/src/JITM.php
Comment (simison) [inline packages/jitm/src/JITM.php]: Empty trailing space in the translation string will likely get accidentally dropped by translators so if it's needed, better use CSS.

File: packages/jitm/src/JITM.php
Comment (simison) [inline packages/jitm/src/JITM.php]: Nit; this and other `target="_blank"` links would need `rel="noreferrer noopener"`

File: packages/jitm/src/JITM.php
Comment (simison) [inline packages/jitm/src/JITM.php]: Would it be possible to leave out `<h2>` in the translation string?

File: packages/jitm/src/JITM.php
Comment (simison) [inline packages/jitm/src/JITM.php]: Since this is un-transpiled ES5 code, you should use `var` instead. (mind IE11)

Comment (simison) [reply packages/jitm/src/JITM.php]: ...tho maybe this bit could be in a `.js` file and enqueued instead so that we catch it better with linters and whatnot. :-)

File: packages/jitm/src/JITM.php
Comment (simison) [reply packages/jitm/src/JITM.php]: As a dev, I'd probably search for something like this from `packages/connection` but not sure how this plays with that package needing to be more generic and not Jetpack-specific?

File: packages/jitm/src/JITM.php
Comment (simison) [inline packages/jitm/src/JITM.php]: You should check if `isset( $_REQUEST['action'] )` first to avoid _ Undefined index_ PHP notices.

File: packages/jitm/src/JITM.php
Comment (simison) [inline packages/jitm/src/JITM.php]: Looks temporary but remember to pass via `__()` + `esc_js`. ;-)

File: packages/jitm/src/JITM.php
Comment (simison) [inline packages/jitm/src/JITM.php]: What do you think if you'd move the link from "this link" to "connect..."?

> If you would like to be the new connection owner for this site, please connect to your WordPress.com account by clicking [this link].
→
> If you would like to be the new connection owner for this site, please [connect to your WordPress.com account].

File: packages/jitm/src/JITM.php
Comment (simison) [inline packages/jitm/src/JITM.php]: "Jetpack servers" sounds very technical, maybe there's a more understandable way to describe the significance of disconnection?

File: packages/jitm/src/JITM.php
Comment (simison) [reply packages/jitm/src/JITM.php]: No idea if this is possible easily but would be nice to have "external" icon after `_blank` links.

File: packages/jitm/src/JITM.php
Comment (simison) [inline packages/jitm/src/JITM.php]: Instead of adding `get_connection_owner()` method, would looping `$user_ids_to_delete` using existing `$connection_manager->is_connection_owner()` method be sufficient enough?

The method sounds fairly useful in a connection package tho. :-)

**Review by simison — COMMENTED** [review]
I really like this idea so I went ahead and left a few minor comments as a review, hope this helps!

Not necessarily in this PR but might make sense to hide the notice when user is deleting themself and WP is stopping them:

Screenshot 2019-08-26 at 09 54 43


Otherwise the notice might confuse.

File: packages/jitm/src/JITM.php
Comment (jeherve) [inline packages/jitm/src/JITM.php]: I think it should be `user` here:

```suggestion
		$user_ids_to_delete        = $_REQUEST['user'];
```

File: packages/jitm/src/JITM.php
Comment (jeherve) [inline packages/jitm/src/JITM.php]: "your site's features", maybe?

File: packages/jitm/src/JITM.php
Comment (jeherve) [inline packages/jitm/src/JITM.php]: I wonder if we should have a different wording when there is only one other admin, the person that's going through the flow?

File: packages/jitm/src/JITM.php
Comment (jeherve) [inline packages/jitm/src/JITM.php]: I wonder if we should be more imperative in this case? It's not just "if you would like", they really need to do this or they'll lose features. The connect button (I think it should be a button) should become the main CTA on the page imo.

**Review by jeherve — CHANGES_REQUESTED** [review]
I like the idea; I think it'd be a great addition to the Users screen. 👍

File: packages/jitm/src/JITM.php
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: We decided at one point to let UI decisions be made by the consumer, and not include UI into the connection package. We can decide the other way later, but it's not important enough to block this PR over IMO.

File: packages/jitm/src/JITM.php
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: `const` looks ok for IE11 except in loops? https://caniuse.com/#feat=const

File: packages/jitm/src/JITM.php
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: "Some Jetpack features may stop working."

File: packages/jitm/src/JITM.php
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: What if the site has 10,000 connected users? (I know unlikely, but this is more performant in scale.)

> The method sounds fairly useful in a connection package tho. :-)

I thought so too and was surprised it didn't already exist.

File: packages/jitm/src/JITM.php
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: Turns out it's both :) 

It's `user` if you are only deleting a single user with the `delete` button straight away, but if you use the checkbox/bulk action, then it's `users` always. I will add a check for `user` as well! good catch.

File: packages/jitm/src/JITM.php
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: Honestly, it's too wordy for me. I think the better way to approach it is to keep it simple with something like "Your site may function properly", and update/link to a support page.

Comment (dereksmart) [reply packages/jitm/src/JITM.php]: fixed!

File: packages/jitm/src/JITM.php
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: Done for all, thanks!

File: packages/jitm/src/JITM.php
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: I updated it anyway

File: packages/jitm/src/JITM.php
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: Yes it's temporary

File: packages/jitm/src/JITM.php
Comment (dereksmart) [reply packages/jitm/src/JITM.php]: Added a check for both!

**Comment by dereksmart** [comment]
I would love if choice of words/design does not block this PR. What do you think about simplifying the wording and instead just linking to a support page such as this where we can add more information? https://jetpack.com/support/primary-user/

File: packages/jitm/src/JITM.php
Comment (simison) [reply packages/jitm/src/JITM.php]: Makes total sense! 👍

**Comment by dereksmart** [comment]
Made some adjustments to the copy and layout. I think it's clearer now, but we'll need to update that support page during beta. 

image

image

Ready for review again.

File: packages/jitm/src/JITM.php
Comment (simison) [reply packages/jitm/src/JITM.php]: Yup 👍  Totally not a blocker, I did mean to mention that. :-) Lots of places in Jetpack seem to miss it already, too.

File: packages/jitm/src/JITM.php
Comment (simison) [reply packages/jitm/src/JITM.php]: IE11 keeps surprising sometimes! 😍

File: packages/jitm/src/JITM.php
Comment (jeherve) [inline packages/jitm/src/JITM.php]: Should it be `$connnected_admins` here, since we are only looking for users that can connect Jetpack to WordPress.com for the site? I was a bit confused at first.

File: packages/jitm/src/JITM.php
Comment (jeherve) [inline packages/jitm/src/JITM.php]: I think using `$connnected_admins` instead of `$connected_users` would be especially useful here. At first I didn't get why we needed to have **more than** 1 connected user. But in fact we need at least 2 connected admins: the current master user and the future one.

File: packages/connection/src/Manager.php
Comment (jeherve) [reply packages/connection/src/Manager.php]: #13323 has now been merged.

File: packages/jitm/src/JITM.php
Comment (jeherve) [inline packages/jitm/src/JITM.php]: Not a blocker, but I think in a future PR we should also mention something about the plan, when the site has one. We want to let folks know that the plan will remain attached the the previous owner.

File: packages/jitm/src/JITM.php
Comment (jeherve) [inline packages/jitm/src/JITM.php]: I think you're missing a texdomain here.

```suggestion
			echo "<a href='" . esc_url( $connect_url ) . "' target='_blank' rel='noopener noreferrer' class='button-primary'>" . esc_html__( 'Connect to WordPress.com', 'jetpack' ) . "</a>";
```

**Review by jeherve — DISMISSED** [review]

File: packages/jitm/src/JITM.php
Comment (jeherve) [inline packages/jitm/src/JITM.php]:
```suggestion
					'include' => array_diff( $connected_admin_ids, array( $connection_owner_id ) ),
```

File: packages/jitm/src/JITM.php
Comment (jeherve) [reply packages/jitm/src/JITM.php]: Only checking for `isset` may not be enough here since `get_userdata( $connection_owner_id )` will always return at least `false`. How about this instead:

```suggestion
			is_a( $connection_owner_userdata, 'WP_User' ) ? esc_html( $connection_owner_userdata->data->user_login ) : ''
```

**Review by jeherve — CHANGES_REQUESTED** [review]

**Review by jeherve — DISMISSED** [review]
This should be good to go like this! 👍

**Review by jeherve — APPROVED** [review]

**Pull request merged**
